// @aindev/connect-sdk/dist/common/firebase.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    Object.defineProperty(o, k2, { enumerable: true, get: function() { return m[k]; } });
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
const app_1 = __importDefault(require("firebase/app"));
require("firebase/auth");
require("firebase/database");
require("firebase/functions");
const constants = __importStar(require("./constants"));
class Firebase {
    constructor(env, config) {
        let firebaseConfig;
        if (config) {
            firebaseConfig = config;
        }
        else {
            firebaseConfig = (env === 'prod')
                ? constants.PROD_FIREBASE_CONFIG : constants.STAGING_FIREBASE_CONFIG;
        }
        this.instance = app_1.default.initializeApp(firebaseConfig);
    }
    getTimestamp() {
        return app_1.default.database.ServerValue.TIMESTAMP;
    }
    getInstance() {
        return this.instance;
    }
    getFunctions() {
        return this.instance.functions();
    }
    getDatabase() {
        return this.instance.database();
    }
}
exports.default = Firebase;
